### ISSUE TITLE
Masking

### ISSUE THREAD

**Issue by gahdritz** [HazyResearch/flash-attention#17]
How difficult would it be to modify the kernels to support arbitrary masks for the attention logits, especially w/ support for broadcasting?

**Comment by tridao** [comment]
We plan to support additive biases (e.g. ALiBi). The bias should ideally take linear memory and not quadratic memory.
Can you explain what shape the arbitrary masks would be? What does tensor broadcasting mean in this case?

Re FP32: We use tensor cores for matrix multiply, which can support fp16 (and bf16 in the near future). The xformers team have some implementation of memory-efficient attention in fp32 ([here](https://github.com/facebookresearch/xformers/pull/267) and [here](https://github.com/facebookresearch/xformers/pull/281)).

**Comment by gahdritz** [comment]
Thanks for the speedy response! 

I'm thinking of applying FlashAttention to our implementation of AlphaFold 2, which has a number of different attention modules with different biases for the pre-softmax quadratic attention matrix `S = Q @ K^T`. To save memory, the biases are deliberately designed to be smaller than the full e.g. [B, H, N, N] `S` tensor---they might have shapes like [1, 1, N, N], or [1, 1, 1, N], and so on, and are only broadcast to `S`'s size at the moment of addition.

BTW, is there a reason the existing `attn_mask` parameter is disabled? Would it be straightforward to get it re-enabled again?

**Comment by tridao** [comment]
Can you point me to the code / paper on how these biases are applied? Are they just added to `S` (potentially with broadcasting)? Are they fixed or trainable (i.e. do we need to compute their gradient in the backward pass)?
 
`attn_mask` is not implemented (yet), it's just there in the Python code to be compatible with our internal testing code.
We do support `key_padding_mask`.

**Comment by gahdritz** [comment]
See e.g. our implementation of what DeepMind calls "[triangular attention](https://github.com/aqlaboratory/openfold/blob/main/openfold/model/triangular_attention.py)". Pseudocode on page 19 [here](https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-021-03819-2/MediaObjects/41586_2021_3819_MOESM1_ESM.pdf). It uses both a fixed mask and a trainable bias.

**Comment by tridao** [comment]
Thanks for the pointers! 
Is the `mask_bias` used to mask out the keys (it should have shape [1, 1, 1, N]), or to mask out arbitrary entries in the attention matrix (shape [1, 1, N, N])?
How large is the sequence length N typically?

**Comment by gahdritz** [comment]
In this particular case, the `mask_bias` masks out the keys. The `triangle_bias`, however, is quadratic. `N` in this case usually ranges from 0-2000, but can go as high as ~10000.

**Comment by gahdritz** [comment]
Sorry I forgot to ask yesterday---do you have an approximate ETA for the `attn_mask` feature? Even without the additive bias feature, FlashAttention would give us a huge speedup in certain modules---we just need that mask.

Thanks!

**Comment by tridao** [comment]
Do you mean a key padding mask (of shape [B, 1, 1, N]), or an arbitrary attention mask of shape [1, 1, N, N]?
We already support key padding mask ([B, 1, 1, N]), query padding mask ([B, 1, N, 1]) and causal mask (for autoregressive modeling).

**Comment by gahdritz** [comment]
I tried playing around with the `key_padding_mask` feature, and as I understand it, it also masks out the corresponding queries and values (it stacks them all and then uses the `unpad_input` function to remove all three at once). Is that incorrect?

**Comment by tridao** [comment]
I've just finished implementing cross-attention where we can support separate query padding and key padding ([here](https://github.com/HazyResearch/flash-attention/blob/5b838a8bef78186196244a4156ec35bbb58c337d/flash_attn/flash_attn_interface.py#L179) and [here](https://github.com/HazyResearch/flash-attention/blob/5b838a8bef78186196244a4156ec35bbb58c337d/flash_attn/flash_attn_interface.py#L213)).
That is, you can all `unpad_input` on the query and then `unpad_input` on the keys and values separately.

I'll push an example usage tomorrow.

**Comment by gahdritz** [comment]
Oh awesome. Thanks!

**Comment by void-main** [comment]
We want customizable masking & biasing as well! Adding these two features would make FlashAttention suitable for a lot more models.

**Comment by OlivierDehaene** [comment]
Hello @tridao, 
Do you still plan on adding support for additive biases? I am currently working on speeding up inference for BLOOM and having the possibility to use flash attention would be great!

**Comment by tridao** [comment]
Yes, it'll be there eventually. I just haven't had as munch bandwidth to work on it recently (conference deadline).

By inference, do you mean text generation? Would Q have seqlen 1 most of the time, due to the `kv_cache`? In that case I think there might not be much benefit to using FlashAttention (attention is not IO-bound).
We have ideas on speeding up text generation in this regime, but that will take a bit of time.

**Comment by guolinke** [comment]
refer to https://github.com/HazyResearch/flash-attention/pull/57

**Comment by zeliu98** [comment]
Hi @tridao, congratulations on your great work!

I have the same issue for using flash attention in Swin-Transformer, especially in shifted window attention. 
It will be appreciated if flash attention can support customizing attention mask and bias.

To be more specific, the shapes in the window attention are:
`q, k, v`: `[batch-size, nheads, ntokens, dim]`
`attn=q@k.T()`: `[batch-size, nheads, ntokens, ntokens]`
`mask`: `[batch-size, 1, ntokens, ntokens]`
`bias`: `[1, nheads, ntokens, ntokens]`

The final attn weights (before softmax) should be: `attn + mask + bias`

**Comment by MayDomine** [comment]
refer to #76

**Comment by calebthomas259** [comment]
Hello @tridao, flash-attention is amazing! Thank you so much for making it!

If possible, I'd also like to request a fully customisable attention bias (i.e. shape `[B, H, N, N]`). This would allow for implementing papers like [Graphormer](https://openreview.net/forum?id=OeWooOxFwDa) (e.g. see Figure 1 of the linked paper).

Thank you for all of your hard work!

**Comment by MayDomine** [comment]
> Hello @tridao, flash-attention is amazing! Thank you so much for making it!
> 
> If possible, I'd also like to request a fully customisable attention bias (i.e. shape `[B, H, N, N]`). This would allow for implementing papers like [Graphormer](https://openreview.net/forum?id=OeWooOxFwDa) (e.g. see Figure 1 of the linked paper).
> 
> Thank you for all of your hard work!

Actually You can  simply use the triton version implmentation to achieve this. But for backward you have to modifiy the origin code .And the performance will decrease because you have to save bias grad in hbm during backward.

**Comment by calebthomas259** [comment]
> > Hello @tridao, flash-attention is amazing! Thank you so much for making it!
> > If possible, I'd also like to request a fully customisable attention bias (i.e. shape `[B, H, N, N]`). This would allow for implementing papers like [Graphormer](https://openreview.net/forum?id=OeWooOxFwDa) (e.g. see Figure 1 of the linked paper).
> > Thank you for all of your hard work!
> 
> Actually You can simply use the triton version implmentation to achieve this. But for backward you have to modify the origin code .And the performance will decrease because you have to save bias grad in hbm during backward.

I did try this at some point, but I was getting errors (I'm not sure whether it was my code being wrong, or just triton bugs). I'll probably try again once triton is more stable

**Comment by PeterL1n** [comment]
Would like to support value_mask for cross attention as well.

**Comment by UCC-team** [comment]
> I've just finished implementing cross-attention where we can support separate query padding and key padding ([here](https://github.com/HazyResearch/flash-attention/blob/5b838a8bef78186196244a4156ec35bbb58c337d/flash_attn/flash_attn_interface.py#L179) and [here](https://github.com/HazyResearch/flash-attention/blob/5b838a8bef78186196244a4156ec35bbb58c337d/flash_attn/flash_attn_interface.py#L213)). That is, you can all `unpad_input` on the query and then `unpad_input` on the keys and values separately.
> 
> I'll push an example usage tomorrow.

Where can we find the example? 
Thank you for all of your hard work!

**Comment by tridao** [comment]
You can look at our BERT [implementation](https://github.com/Dao-AILab/flash-attention/blob/4b661a569dc0e476b07a0ac13a892fc8d120b3b0/flash_attn/models/bert.py#L144).

**Comment by UCC-team** [comment]
> You can look at our BERT [implementation](https://github.com/Dao-AILab/flash-attention/blob/4b661a569dc0e476b07a0ac13a892fc8d120b3b0/flash_attn/models/bert.py#L144).


If convert a sparse mask to cu_seqlens and max_seqlen using unpad_input, we will get incorrect results. The results of `attention_ref(q, k, v, q_mask, k_mask)` and `flash_attn_varlen_func(q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k)` do not match.
(The sparse mask is not a padding mask; it might appear in the middle rather than the end.)
```
_, _, cu_seqlens_q, max_seqlen_q = unpad_input(q, q_mask)
_, _, cu_seqlens_k, max_seqlen_k = unpad_input(k, k_mask)
```
How should we solve this problem？Thanks!

**Comment by tridao** [comment]
If the mask is in the middle of the sequence, that's not supported right now.

**Comment by UCC-team** [comment]
> If the mask is in the middle of the sequence, that's not supported right now.

Can the `blockmask` be used as any mask mentioned above in the `flash_blocksparse_attn_func`  interface?
[flash_blocksparse_attn_func](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/flash_blocksparse_attn_interface.py#L188)
